pandas/core/frame.py

```diff
@@ -4302,7 +4303,29 @@ def set_index(
         if not inplace:
             return frame
 
+    @overload
     def reset_index(
+        self,
+        level: Optional[Union[Hashable, Sequence[Hashable]]] = None,
+        drop: bool = False,
+        *,
```

Comment:
Is it possible to combine Literal and overload here instead of this? I don’t think this is 100% accurate if a user provides `inplace=False`

Reply:
we can't use Literal til Python 3.8 or unless we adopt typing extensions.

If we pass `inplace=False` the return type will match the second signature and so the return type will be the Union. This is not inaccurate and no change from the status quo.

If we don't pass an `inplace` argument, the return type will be DataFrame and additional asserts, casts or ignores aren't needed. This is an improvement from the status quo.

Reply:
Whether explicitly passing `inplace=False` or just sticking with the default signature the expected return type shouldn't be different, which is what is being introduced here. The ideal overload should be something like "return None if inplace=True, otherwise return a DataFrame".

While the current approach may appease most type checking it introduces unexpected behavior when providing the inplace keyword; would rather try to integrate a Literal type or actually deprecate inplace.